File: .claude/skills/threads-post/scripts/publisher.js
#!/usr/bin/env node
/**
 * threads_publisher.js — Pubblica post su Threads via Meta Graph API
 * Uso:
 *   node publisher.js --dry-run
 *   node publisher.js --text "testo" [--image-url URL]
 */

import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';
import { config } from 'dotenv';

const __dirname = dirname(fileURLToPath(import.meta.url));
config({ path: resolve(__dirname, '../../../../.env') });

const ACCESS_TOKEN = process.env.THREADS_ACCESS_TOKEN;
const USER_ID = process.env.THREADS_USER_ID;
const API_BASE = 'https://graph.threads.net/v1.0';

function checkCredentials() {
  if (!ACCESS_TOKEN || !USER_ID) {
    console.error('❌ Credenziali Threads mancanti. Compila THREADS_ACCESS_TOKEN e THREADS_USER_ID in .env');
    console.error('   Guida: .claude/setup/api-credentials.md');
    process.exit(1);
  }
}

async function dryRun() {
  checkCredentials();
  const url = new URL(`${API_BASE}/${USER_ID}`);
  url.searchParams.set('fields', 'id,username');
  url.searchParams.set('access_token', ACCESS_TOKEN);
  const r = await fetch(url);
  if (r.ok) {
    const data = await r.json();
    console.log(`✓ Auth OK — dry run completato. Account: @${data.username ?? 'N/A'}`);
  } else {
    console.error(`❌ Auth fallita: ${r.status} ${await r.text()}`);
    process.exit(1);
  }
}

async function createMediaContainer(text, imageUrl = null) {
  const url = new URL(`${API_BASE}/${USER_ID}/threads`);
  url.searchParams.set('media_type', imageUrl ? 'IMAGE' : 'TEXT');
  url.searchParams.set('text', text);
  url.searchParams.set('access_token', ACCESS_TOKEN);
  if (imageUrl) url.searchParams.set('image_url', imageUrl);

  const r = await fetch(url, { method: 'POST' });
  if (!r.ok) throw new Error(`createMediaContainer fallito: ${r.status} ${await r.text()}`);
  const data = await r.json();
  console.log(`✓ Container creato: ${data.id}`);
  return data.id;
}

async function publishContainer(containerId) {
  await new Promise(res => setTimeout(res, 3000));
  const url = new URL(`${API_BASE}/${USER_ID}/threads_publish`);
  url.searchParams.set('creation_id', containerId);
  url.searchParams.set('access_token', ACCESS_TOKEN);

  const r = await fetch(url, { method: 'POST' });
  if (!r.ok) throw new Error(`publishContainer fallito: ${r.status} ${await r.text()}`);
  const data = await r.json();
  const postUrl = `https://www.threads.net/@giacomomontepino/post/${data.id}`;
  console.log(`✓ Post pubblicato: ${postUrl}`);
  return postUrl;
}

async function main(text, imageUrl = null) {
  checkCredentials();
  const containerId = await createMediaContainer(text, imageUrl);
  return publishContainer(containerId);
}

// CLI
const args = process.argv.slice(2);
const get = (flag) => { const i = args.indexOf(flag); return i !== -1 ? args[i + 1] : null; };

if (args.includes('--dry-run')) {
  dryRun().catch(e => { console.error(e.message); process.exit(1); });
} else if (get('--text')) {
  main(get('--text'), get('--image-url')).catch(e => { console.error(e.message); process.exit(1); });
} else {
  console.log('Uso: node publisher.js --dry-run | --text "testo" [--image-url URL]');
}

File: .claude/skills/threads-post/scripts/publisher.py
#!/usr/bin/env python3
"""
threads_publisher.py — Pubblica post su Threads via Meta Graph API
Uso:
  publisher.py --dry-run                           Testa autenticazione
  publisher.py --text "testo" [--image-url URL]    Pubblica post
"""

import argparse
import os
import sys
import time
from pathlib import Path

import requests
from dotenv import load_dotenv

load_dotenv(Path(__file__).parents[5] / ".env")

ACCESS_TOKEN = os.getenv("THREADS_ACCESS_TOKEN")
USER_ID = os.getenv("THREADS_USER_ID")
API_BASE = "https://graph.threads.net/v1.0"


def _check_credentials():
    if not ACCESS_TOKEN or not USER_ID:
        print("❌ Credenziali Threads mancanti. Compila THREADS_ACCESS_TOKEN e THREADS_USER_ID in .env", file=sys.stderr)
        print("   Guida: .claude/setup/api-credentials.md", file=sys.stderr)
        sys.exit(1)


def dry_run():
    _check_credentials()
    r = requests.get(
        f"{API_BASE}/{USER_ID}",
        params={"fields": "id,username", "access_token": ACCESS_TOKEN},
    )
    if r.status_code == 200:
        data = r.json()
        print(f"✓ Auth OK — dry run completato. Account: @{data.get('username', 'N/A')}")
    else:
        print(f"❌ Auth fallita: {r.status_code} {r.text}", file=sys.stderr)
        sys.exit(1)


def create_media_container(text, image_url=None):
    params = {
        "media_type": "IMAGE" if image_url else "TEXT",
        "text": text,
        "access_token": ACCESS_TOKEN,
    }
    if image_url:
        params["image_url"] = image_url

    r = requests.post(f"{API_BASE}/{USER_ID}/threads", params=params)
    r.raise_for_status()
    container_id = r.json()["id"]
    print(f"✓ Container creato: {container_id}")
    return container_id


def publish_container(container_id):
    # Attendi che il container sia pronto (Media Container può richiedere qualche secondo)
    time.sleep(3)
    r = requests.post(
        f"{API_BASE}/{USER_ID}/threads_publish",
        params={"creation_id": container_id, "access_token": ACCESS_TOKEN},
    )
    r.raise_for_status()
    post_id = r.json()["id"]
    post_url = f"https://www.threads.net/@giacomomontepino/post/{post_id}"
    print(f"✓ Post pubblicato: {post_url}")
    return post_url


def main(text, image_url=None):
    _check_credentials()
    container_id = create_media_container(text, image_url=image_url)
    return publish_container(container_id)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Threads Publisher")
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--text", type=str)
    parser.add_argument("--image-url", type=str)
    args = parser.parse_args()

    if args.dry_run:
        dry_run()
    elif args.text:
        main(args.text, image_url=args.image_url)
    else:
        parser.print_help()
